src/sentry/coreapi.py

```diff
@@ -423,13 +431,28 @@ def validate_data(self, project, data):
         if 'timestamp' in data:
             try:
                 self._process_data_timestamp(data)
+            except FutureInvalidTimestamp as e:
+                data['errors'].append({
+                    'type': EventError.INVALID_TIMESTAMP_SKEW,
+                    'value': data['timestamp'],
+                    'reason': 'future',
+                    'delta': e.message,
+                })
+                del data['timestamp']
+            except PastInvalidTimestamp as e:
+                data['errors'].append({
+                    'type': EventError.INVALID_TIMESTAMP_SKEW,
+                    'value': data['timestamp'],
+                    'reason': 'past',
+                    'delta': e.message,
+                })
+                del data['timestamp']
             except InvalidTimestamp as e:
                 self.log.debug(
```

Comment:
> this is the only one logging

I explicitly didn't think it was necessary to log for the cases with clock skew because this isn't valuable information to us, as admins of Sentry. It's only valuable information to the end user which is being surfaced anyways. This is different than the case of a flat out malformed timestamp value that may be more useful to us developing against Sentry, etc.

> also needs tests

There are some that cover all of these cases. :)

> also why not re-use same error type and have a diff reason on each one?

I originally did, but I preferred to pick out the "skew" case, again, similar reason to logging. The fact that the clock is skewed is more significant and common than the fact that the timestamp is entirely wrong. Differentiating these two cases allows a better error message to be shown.